# Sash Workflow Overview

## Summary

The **Sash Workflow** comprises three primary pipelines: **Somatic Small Variants**, **Somatic Structural Variants**, and **Germline Variants**. These pipelines utilize **Bolt**, a Python package designed for modular processing, and leverage outputs from the **DRAGEN Variant Caller** alongside **HMFtools in Oncoanalyser**. Each pipeline is tailored to a specific type of genomic variant, incorporating filtering, annotation, and HTML reports for research and curation.

## Workflows

### Somatic Small Variants

#### General

In the **Somatic Small Variants** workflow, variant detection is performed using the **DRAGEN Variant Caller** and **Oncoanalyser (SAGE, Purple)** outputs. Itâ€™s structured into four steps: **Rescue**, **Annotation**, **Filter**, and **Report**. The final outputs include an **HTML report** summarizing the results.

#### Summary

1. **Rescue** variants using SAGE to recover low-frequency alterations in clinically important hotspots.
2. **Annotate** variants with clinical and functional information using PCGR.
3. **Filter** variants based on allele frequency (AF), supporting reads (AD), and population frequency (gnomAD AF), removing low-confidence and common variants.
4. **Report** final annotated variants in a comprehensive HTML format.

### Details

#### BOLT_SMLV_SOMATIC_RESCUE

<details markdown="1">
<summary>Output files</summary>

- `output/`
  - `output/${meta.tumor_id}.rescued.vcf.gz`: Rescued somatic VCF file.
  - `output/${meta.tumor_id}.rescued.vcf.gz.tbi`: Index file for the rescued VCF.

</details>

The `BOLT_SMLV_SOMATIC_RESCUE` process rescues somatic variants using the BOLT tool. The output includes the rescued VCF file and its index.

#### BOLT_SMLV_SOMATIC_ANNOTATE

<details markdown="1">
<summary>Output files</summary>

- `output/`
  - `output/${meta.tumor_id}.annotations.vcf.gz`: Annotated somatic VCF file.
  - `output/${meta.tumor_id}.annotations.vcf.gz.tbi`: Index file for the annotated VCF.

</details>

The `BOLT_SMLV_SOMATIC_ANNOTATE` process annotates somatic variants using the BOLT tool. The output includes the annotated VCF file and its index.

#### BOLT_SMLV_SOMATIC_FILTER

<details markdown="1">
<summary>Output files</summary>

- `output/`
  - `output/${meta.tumor_id}*pass.vcf.gz`: Filtered somatic VCF file.
  - `output/${meta.tumor_id}*pass.vcf.gz.tbi`: Index file for the filtered VCF.
  - `output/${meta.tumor_id}*filters_set.vcf.gz`: VCF file with filters set.

</details>

The `BOLT_SMLV_SOMATIC_FILTER` process filters somatic variants using the BOLT tool. The output includes the filtered VCF file, its index, and the VCF file with filters set.

#### PAVE_SOMATIC

<details markdown="1">
<summary>Output files</summary>

- `output/`
  - `output/${meta.tumor_id}.pave.vcf.gz`: PAVE somatic VCF file.
  - `output/${meta.tumor_id}.pave.vcf.gz.tbi`: Index file for the PAVE VCF.

</details>

The `PAVE_SOMATIC` process processes somatic variants using the PAVE tool. The output includes the PAVE VCF file and its index.

### Sash Module Outputs

**1. Somatic SNVs:**

- `smlv_somatic/filter/{tid}.pass.vcf.gz`: Contains somatic single nucleotide variants (SNVs) with filtering applied.

**2. Somatic SVs:**

- `sv_somatic/prioritise/{tid}.sv.prioritised.vcf.gz`: Contains somatic structural variants (SVs) with prioritization applied.

**3. Somatic CNVs:**

- `cancer_report/cancer_report_tables/purple/{sid}_{tid}-purple_cnv_som.tsv.gz`: Contains somatic copy number variations (CNVs) data.

**4. Somatic Gene CNVs:**

- `cancer_report/cancer_report_tables/purple/{sid}_{tid}-purple_cnv_som_gene.tsv.gz`: Contains gene-level somatic copy number variations (CNVs) data.

**5. Germline SNVs:**

- `dragen_germline_output/{nid}.hard-filtered.vcf.gz`: Contains germline single nucleotide variants (SNVs) with hard filtering applied.

**6. Purple Purity, Ploidy, MS Status:**

- `purple/{tid}.purple.purity.tsv`: Contains estimated tumor purity, ploidy, and microsatellite status.

**7. PCGR JSON with TMB:**

- `smlv_somatic/report/pcgr/{tid}.pcgr_acmg.grch38.json.gz`: Contains PCGR annotations, including tumor mutational burden (TMB).

**8. DRAGEN HRD Score:**

- `dragen_somatic_output/{tid}.hrdscore.tsv`: Contains homologous recombination deficiency (HRD) score from DRAGEN analysis.

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.

## Conclusion

This document provides an overview of the output files generated by the pipeline. For more detailed information about each step and the tools used, please refer to the respective documentation and help pages.